File: Basics/C#/CSharp.Basics/Samples/Extensions/AdoptionService.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CSharp.Basics.Samples.Extensions
{
    public class AdoptionService
    {
        public void AdoptDog(AdoptableDog dog)
        {
            dog.Adopt();
        }
    }

    public static class AdoptableExtensions
    {
        public static void Adopt(this AdoptableDog dog)
        {
            dog.IsAdopted = true;
        }
    }

    public class AdoptableDog
    {
        public bool IsAdopted { get; set; }
    }
}

File: Basics/Kotlin/app/src/main/java/com/suavepirate/kotlinbasics/samples/extensions/AdoptionService.kt
package com.suavepirate.kotlinbasics.samples.extensions

class AdoptionService {
    fun adoptDog(dog: AdoptableDog) {
        dog.adopt()
    }
}

// extension functions
fun AdoptableDog.adopt() {
    this.isAdopted = true
}

// extension properties
val AdoptableDog.name : String
    get() = "${this.firstName} ${this.lastName}"

// companion object extensions

fun AdoptableDog.Companion.getShelters() {
    // ...
}

data class AdoptableDog(var firstName: String, var lastName: String, var isAdopted: Boolean) {
    companion object {}
}